File: merge.rb
def merge_sort(array)
  return array if array.length < 2
  left_half = array[0...array.length/2]
  right_half = array[array.length/2...array.length]
  sorted_left = merge_sort(left_half)
  sorted_right = merge_sort(right_half)
  merge(sorted_left, sorted_right)
end

def merge(left, right)
  sorted_array = []
  until left.empty? || right.empty?
      sorted_array << (left.first <= right.first ? left.shift : right.shift)
  end
  sorted_array + left + right
end

array = [38, 27, -43, 3, -9, 82, -10, 19]
p "Unsorted array: ", array
sorted_array = merge_sort(array)
p "Sorted array: ", sorted_array

File: sorting/hard/merge.js
const mergeSort = (a) => {
    if(a.length < 2) return a
    let left = a.slice(0,a.length/2)
    let right = a.slice(a.length/2, a.length)
    let sortedLeft = mergeSort(left)
    let sortedRight = mergeSort(right)
    return merge(sortedLeft,sortedRight)
}

const merge = (left,right) => {
    let sorted = []
    while(left.length > 0 && right.length > 0){
        left[0] <= right[0] ? sorted.push(left.shift()) : sorted.push(right.shift())
    }
    return sorted.concat(left,right)
}

a = [7,-1,5,-11,60,2,1,8]
console.log("Unsorted Array: ",a);
console.log("Sorted Array: ",mergeSort(a))